### DISCUSSION TITLE
Questions about training

### DISCUSSION THREAD

**Discussion by asluborski** [ultralytics/yolov5#7187] [Q&A]
Hi I have 3 questions about training

1.) I have dataset, there exists some annotation .txt files without images. Is it safe to ignore these or will yolo read them and get confused? I was able to start succesfully training for 2 epoch and it did not complain but I am not sure if yolo is ignoring it or if it is making noise in the network.

2.) When tuning/retraining model is it good practice to use --img-weights argument to help train to detect the objects with lower accuracy? 

3.) From what I understand, Yolo performs augmentation of images during training. Is this done on the image or are copies made of the image, hence increasing the size of the dataset? I ask this because I have existing augmented datasets. 

Thank you, getting these answers would be extremely helpful!

**Comment by glenn-jocher** [comment]
@asluborski 👋 Hello! Thanks for asking about **image augmentation**. YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

### Augmentation Hyperparameters

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch-low.yaml
```
https://github.com/ultralytics/yolov5/blob/b94b59e199047aa8bf2cdd4401ae9f5f42b929e6/data/hyps/hyp.scratch-low.yaml#L6-L34

### Augmentation Previews

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
<img src="https://user-images.githubusercontent.com/26833433/131255960-b536647f-7c61-4f60-bbc5-cb2544d71b2a.jpg" width="800">


### YOLOv5 Albumentations Integration

[YOLOv5](https://github.com/ultralytics/yolov5) 🚀 is now fully integrated with [Albumentations](https://github.com/albumentations-team/albumentations), a popular open-source image augmentation package. Now you can train the world's best Vision AI models even better with custom Albumentations 😃! 

PR https://github.com/ultralytics/yolov5/pull/3882 implements this integration, which will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.3` is installed in your environment. See https://github.com/ultralytics/yolov5/pull/3882 for full details.

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img src="https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg" width="800">

Good luck 🍀 and let us know if you have any other questions!

**Comment by Abhiramborige** [comment]
> 

Hey, the 2nd question as posted by you, how to actually proceed with tuning the training for classes (objects) with less accuracy/confidence level?
Could you please guide me in the same.

**Comment by Abhiramborige** [comment]
@glenn-jocher I have to train the model to detect small icons of size upto 80px on a background of 2500px height and width. 
These are the parameters I use in the configurations.
# configurations.yaml
lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.1  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.3  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 0.7  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
# anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)

Could you please guide me on whether the configurations are all good to go for these kind of detections to be successful ?
or any changes i have to make it here so as the model detects in better way.

I am using this command for training:
python --optimizer Adam --imgsz 960 --epochs 200 --patience 100 --seed 32 --data data.yaml --cfg yolov5s.yaml --weights '' --batch-size 5 --device 0 --hyp configurations.yaml
Thankyou.
